cpp/src/io/avro/reader_impl.cu

```diff
@@ -329,76 +339,73 @@ void reader::impl::decode_data(const rmm::device_buffer& block_data,
 
   for (size_t i = 0; i < out_buffers.size(); i++) {
     const auto col_idx          = selection[i].first;
-    const auto schema_null_idx  = _metadata->columns[col_idx].schema_null_idx;
+    const auto schema_null_idx  = meta->columns[col_idx].schema_null_idx;
     out_buffers[i].null_count() = (schema_null_idx >= 0) ? schema_desc[schema_null_idx].count : 0;
   }
 }
 
-reader::impl::impl(std::unique_ptr<datasource> source,
-                   avro_reader_options const& options,
-                   rmm::mr::device_memory_resource* mr)
-  : _mr(mr), _source(std::move(source)), _columns(options.get_columns())
-{
-  // Open the source Avro dataset metadata
-  _metadata = std::make_unique<metadata>(_source.get());
-}
-
-table_with_metadata reader::impl::read(avro_reader_options const& options,
-                                       rmm::cuda_stream_view stream)
+table_with_metadata read_avro(std::unique_ptr<cudf::io::datasource>&& source,
+                              avro_reader_options const& options,
+                              rmm::cuda_stream_view stream,
+                              rmm::mr::device_memory_resource* mr)
 {
   auto skip_rows = options.get_skip_rows();
   auto num_rows  = options.get_num_rows();
   num_rows       = (num_rows != 0) ? num_rows : -1;
   std::vector<std::unique_ptr<column>> out_columns;
   table_metadata metadata_out;
 
+  // Open the source Avro dataset metadata
+  auto meta = std::make_unique<metadata>(source.get());
```

Comment:
What? You are creating a `unique_ptr` from the content of another `unique_ptr`? This seems not right, and you may encounter undefined behavior. If you want to access the `source` variable through an instance of `metadata` (I assume that `metadata` and `datasource` are on the same class hierrachy), please use `dynamic_cast`:
```suggestion
  auto const meta = dynamic_cast<metadata>(source.get());
```